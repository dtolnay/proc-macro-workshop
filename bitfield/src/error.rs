use std::fmt::{self, Display};

#[derive(Debug)]
pub struct Error {
    typename: &'static str,
    value: u64,
    width: u8,
}

impl Error {
    #[doc(hidden)]
    pub fn new(typename: &'static str, value: u64, width: u8) -> Error {
        Error {
            typename,
            value,
            width,
        }
    }

    pub fn raw_value(&self) -> u64 {
        self.value
    }
}

impl Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "unrecognized bit pattern for enum {}: 0b{:02$b}",
            self.typename, self.value, self.width as usize,
        )
    }
}

impl std::error::Error for Error {}
