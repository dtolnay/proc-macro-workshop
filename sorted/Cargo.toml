[package]
name = "sorted"
version = "0.0.0"
autotests = false
edition = "2021"
publish = false

[lib]
proc-macro = true

[[test]]
name = "tests"
path = "tests/progress.rs"

[dev-dependencies]
trybuild = { version = "1.0.49", features = ["diff"] }

[dependencies]
proc-macro2 = "1.0"
quote = "1.0"
syn = { version = "2.0", features = ["full", "visit-mut"] }
