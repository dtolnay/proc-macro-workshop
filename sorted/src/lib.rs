mod check;
mod compare;
mod emit;
mod format;
mod parse;
mod visit;

use proc_macro::TokenStream;
use quote::quote;
use syn::{parse_macro_input, ItemFn};

use crate::emit::emit;
use crate::parse::{Input, Nothing};

#[proc_macro_attribute]
pub fn sorted(args: TokenStream, input: TokenStream) -> TokenStream {
    let original = input.clone();

    let _ = parse_macro_input!(args as Nothing);
    let input = parse_macro_input!(input as Input);
    let kind = input.kind();

    match check::sorted(input) {
        Ok(()) => original,
        Err(err) => emit(err, kind, original),
    }
}

#[proc_macro_attribute]
pub fn check(args: TokenStream, input: TokenStream) -> TokenStream {
    let _ = parse_macro_input!(args as Nothing);
    let mut input = parse_macro_input!(input as ItemFn);

    visit::check(&mut input);

    TokenStream::from(quote!(#input))
}
