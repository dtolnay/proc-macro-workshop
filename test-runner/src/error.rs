use std::env;
use std::fmt::{self, Display};
use std::io;
use std::path::PathBuf;

#[remain::sorted]
pub enum Error {
    Cargo(io::Error),
    CargoFail,
    Io(io::Error),
    Mismatch,
    Open(PathBuf, io::Error),
    PkgName(env::VarError),
    ProjectDir,
    ReadStderr(io::Error),
    RunFailed,
    ShouldNotHaveCompiled,
    Toml(toml::ser::Error),
    WriteStderr(io::Error),
}

pub type Result<T> = std::result::Result<T, Error>;

impl Display for Error {
    #[remain::check]
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        use self::Error::*;

        #[sorted]
        match self {
            Cargo(e) => write!(f, "failed to execute cargo: {}", e),
            CargoFail => write!(f, "cargo reported an error"),
            Io(e) => write!(f, "{}", e),
            Mismatch => write!(f, "compiler error does not match expected error"),
            Open(path, e) => write!(f, "{}: {}", path.display(), e),
            PkgName(e) => write!(f, "failed to detect CARGO_PKG_NAME: {}", e),
            ProjectDir => write!(f, "failed to determine name of project dir"),
            ReadStderr(e) => write!(f, "failed to read stderr file: {}", e),
            RunFailed => write!(f, "execution of the test case was unsuccessful"),
            ShouldNotHaveCompiled => {
                write!(f, "expected test case to fail to compile, but it succeeded")
            }
            Toml(e) => write!(f, "{}", e),
            WriteStderr(e) => write!(f, "failed to write stderr file: {}", e),
        }
    }
}

impl Error {
    pub fn already_printed(&self) -> bool {
        use self::Error::*;

        match self {
            CargoFail | Mismatch | RunFailed | ShouldNotHaveCompiled => true,
            _ => false,
        }
    }
}

impl From<io::Error> for Error {
    fn from(err: io::Error) -> Self {
        Error::Io(err)
    }
}

impl From<toml::ser::Error> for Error {
    fn from(err: toml::ser::Error) -> Self {
        Error::Toml(err)
    }
}
