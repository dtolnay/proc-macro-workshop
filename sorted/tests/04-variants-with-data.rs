// This test is similar to the previous where want to ensure that the macro
// correctly generates an error when the input enum is out of order, but this
// time it is using an enum that also has data associated with each variant.

use sorted::sorted;

use std::env::VarError;
use std::error::Error as StdError;
use std::fmt;
use std::io;
use std::str::Utf8Error;

#[sorted]
pub enum Error {
    Fmt(fmt::Error),
    Io(io::Error),
    Utf8(Utf8Error),
    Var(VarError),
    Dyn(Box<dyn StdError>),
}

fn main() {}
