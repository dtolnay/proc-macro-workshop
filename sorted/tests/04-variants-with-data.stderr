error: Dyn should sort before Fmt
   |
19 |     Dyn(Box<dyn StdError>),
   |     ^^^
